include/cadmium/core/modeling/coupled.hpp

```diff
@@ -270,6 +272,126 @@ namespace cadmium {
 		}
+
+
+		void flatten(){
+        	std::vector<std::shared_ptr<Coupled>> toFlatten;
+        	for (auto& component: components) {
+        		auto coupled = std::dynamic_pointer_cast<Coupled>(component);
+        		if (coupled != nullptr) {
+        			toFlatten.push_back(coupled);
+        		}
+        	}
+
+        	for(auto itc = toFlatten.begin(); itc != toFlatten.end(); itc++){
+        		(*itc)->flatten();
+        		removePortsAndCouplings((*itc));
+        		for(auto itcomp = components.begin(); itcomp != components.end(); itcomp++){
+        			if(*itc == *itcomp) {
+        				components.erase(itcomp);
+        			}
+        		}
+        	}
+
+        	toFlatten.clear();
+
+        	if(parent != nullptr) {
+        		std::vector<std::shared_ptr<PortInterface>> leftBridgeEIC = createLeftBridge(parent->getEICs());
+        		std::vector<std::shared_ptr<PortInterface>> leftBridgeIC = createLeftBridge(parent->getICs());
+
+        		std::vector<std::shared_ptr<PortInterface>> rightBridgeEOC = createRightBridge(parent->getEOCs());
+        		std::vector<std::shared_ptr<PortInterface>> rightBridgeIC = createRightBridge(parent->getICs());
+
+        		completeLeftBridge(EIC, leftBridgeEIC, parent->getEICs());
+        		completeLeftBridge(EIC, leftBridgeIC, parent->getICs());
+        		completeRightBridge(EOC, rightBridgeEOC, parent->getEOCs());
+        		completeRightBridge(EOC, rightBridgeIC, parent->getICs());
+
+            	for (auto& component: components) {
+            		parent->addComponent(component);
+            	}
+
+            	for (auto ic: IC) {
+            		parent->getICs().push_back(ic);
+            	}
+        	}
+		}
+
+        void removePortsAndCouplings(std::shared_ptr<Coupled> child) {
+        	std::vector<std::shared_ptr<PortInterface>> inPorts = child->getInPorts();
+        	for(auto it = inPorts.begin(); it != inPorts.end(); it++){
+        		for(auto itc = EIC.begin(); itc != EIC.end(); itc++){
+        			if(*it == std::get<1>(*itc)) {
+            			EIC.erase(itc);
+            		}
+        		}
+        		for(auto itc = IC.begin(); itc != IC.end(); itc++){
+        			if(*it == std::get<1>(*itc)) {
+            			IC.erase(itc);
+            		}
+        		}
+        	}
+        	std::vector<std::shared_ptr<PortInterface>> outPorts = child->getOutPorts();
+        	for(auto it = outPorts.begin(); it != outPorts.end(); it++){
+        		for(auto itc = EOC.begin(); itc != EOC.end(); itc++){
+        			if(*it == std::get<0>(*itc)) {
+        				EOC.erase(itc);
+        			}
+        		}
+        		for(auto itc = IC.begin(); itc != IC.end(); itc++){
+        			if(*it == std::get<0>(*itc)) {
+        				IC.erase(itc);
+        			}
+        		}
+        	}
+        }
+
+        std::vector<std::shared_ptr<PortInterface>> createLeftBridge(std::vector<coupling> couplings) {
+        	std::vector<std::shared_ptr<PortInterface>> leftBridge;
+            for (auto iPort : this->getInPorts()) {
```

Comment:
In this kind of iterators, remember always to use `auto& x`. It will be a little bit more efficient

Reply:
Done